misc/python/materialize/mzbuild.py

```diff
@@ -190,6 +193,69 @@ def inputs(self) -> Set[str]:
         return set(git.expand_globs(self.rd.root, f"{self.source}/{self.matching}"))
 
 
+class S3Upload(PreImage):
+    """A `PreImage` action which uploads an executable to S3.
+
+    If no AWS credentials are configured in the environment, no action is taken.
+
+    Otherwise, the ELF file with the specified `NAME` is scanned to get its build ID,
+    then it is uploaded to the path `buildid/BUILDID/executable` in the specified bucket.
+
+    If an ELF file called `NAME.debug` exists, and if its build ID matches
+    that of the main executable, it is uploaded to the path `buildid/BUILDID/debuginfo`.
+    """
```

Comment:
This looks super specific to uploading debuginfo—seems like it wants a more specific name than `S3Upload`.

```suggestion
class S3UploadDebuginfo(PreImage):
    """A `PreImage` action which uploads an executable and its debug info to S3.

    If no AWS credentials are configured in the environment, no action is taken.

    Otherwise, the ELF file with the specified `NAME` is scanned to get its build ID,
    then it is uploaded to the path `buildid/BUILDID/executable` in the specified bucket.

    If an ELF file called `NAME.debug` exists, and if its build ID matches
    that of the main executable, it is uploaded to the path `buildid/BUILDID/debuginfo`.
    """
```

Reply:
I didn't intend for it to be specific to debuginfo -- one can imagine a world where someone is interested in archiving just executables in s3 indexed by build ID, even though we aren't currently interested in that today.

Reply:
(That's why I just return gracefully if the debuginfo isn't found or doesn't match, rather than throwing an exception)

Reply:
Well it's still not general purpose enough to be called `S3Upload`, IMO! And not erroring if the debuginfo isn't found seems likely to mask bugs when our only purpose for this is to upload the debuginfo.